### ISSUE TITLE
Chrome Cannot Find Iconset

### ISSUE THREAD

**Issue by jatsrt** [PolymerElements/iron-icon#19]
Seeing this error in the console only on chrome.

iron-icon.html:167 [iron-icon::_updateIcon]: could not find iconset `icons`, did you import the iconset?

This happens for all iconsets that are included and I try to use.  I also tried changing the import order of iron-icons.html to before and after the iconsets are imported and nothing changes.

Icons from the icon sets show up fine in FABs and other items.

Chrome:
![screen shot 2015-05-30 at 5 09 26 pm](https://cloud.githubusercontent.com/assets/635242/7899384/c7f82796-06ee-11e5-936c-5e4c3d03ed80.png)

Safari osx and iOS:
![screen shot 2015-05-30 at 5 09 46 pm](https://cloud.githubusercontent.com/assets/635242/7899385/c7f8f7e8-06ee-11e5-9992-42d44c2266f7.png)

**Comment by mattiLeBlanc** [comment]
Yes I have the same issue: [iron-icon::_updateIcon]: could not find iconset `social`, did you import the iconset?

I am migrating from version 0.5 to 1.0 and running into a lot of issues.

**Comment by sjmiles** [comment]
Root cause believed to be https://github.com/PolymerElements/iron-iconset-svg/issues/10

**Comment by otroboe** [comment]
Same issue here :(
Works on FF 38, but fails on Chrome 43.

**Comment by emmanuelbuah** [comment]
Same here

**Comment by rohanray** [comment]
I am facing the same issue: https://github.com/PolymerElements/iron-icons/issues/14

**Comment by jurgenfink** [comment]
Found the solution.

The cause of the warning in Chrome debugger is due to the wrong timing of loading the link imports in the right sequence.

**Solution:**

 **1.) Remove link imports in the iron-icons** (and other icon-sets if needed, like maps, social, etc ...):
-  public
  -  bower_components
    - iron-icons
      -  **iron-icons.html**
      -  **maps-icons.html** (optional, if you are using them)
      -  **social-icons.html** (optional, if you are using them)

**iron-icons.html:**

before:

```
<!--@group Iron Elements
@element iron-icons
@demo demo/index.html
-->
<link rel='import' href='../iron-icon/iron-icon.html'>
<link rel='import' href='../iron-iconset-svg/iron-iconset-svg.html'>

<iron-iconset-svg name="icons" size="24">
    <svg><defs>
    <g id="3d-rotation"><path d="M7.52 21.48C ..... etc ..... /></g>
    </defs></svg>
</iron-iconset-svg>
```

after:

```
<!--@group Iron Elements
@element iron-icons
@demo demo/index.html
-->
<!--<link rel='import' href='../iron-icon/iron-icon.html'>
<link rel='import' href='../iron-iconset-svg/iron-iconset-svg.html'>-->

<iron-iconset-svg name="icons" size="24">
    <svg><defs>
    <g id="3d-rotation"><path d="M7.52 21.48C ..... etc ..... /></g>
    </defs></svg>
</iron-iconset-svg>
```

The initial link-imports (dependencies) are blocking (not loading async but sync, which is good because that's the way it should be). 
However, within the code of `<link rel='import' href='../iron-icon/iron-icon.html'>`  iron-icon is making reference to the icon set that could not load yet (`<iron-iconset-svg name="icons" size="24">` etc ...) because it comes after that initial link-import (due to blocking nature of link import). Hence, at line 164 it cannot find the Default iconset **icons**, and therefore throwing the famous warning at line 167:

> could not find iconset icons, did you import the iconset?

**2.) Load required dependencies** in your project file in the correct sequence:

```
<head>
  <meta charset="utf-8" />
  <title></title>

  <script src="/bower_components/webcomponentsjs/webcomponents-lite.min.js"></script>
  <link rel="import" href="/bower_components/polymer/polymer.html">

  <link rel="import" href="/bower_components/iron-meta/iron-meta.html">
  <link rel="import" href="/bower_components/iron-flex-layout/iron-flex-layout.html">
  <link rel="import" href="/bower_components/iron-iconset-svg/iron-iconset-svg.html">
  <link rel="import" href="/bower_components/iron-iconset/iron-iconset.html">

  <link rel="import" href="/bower_components/iron-icons/iron-icons.html">
  <link rel="import" href="/bower_components/iron-icons/maps-icons.html">
  <link rel="import" href="/bower_components/iron-icons/social-icons.html">

  <link rel="import" href="/bower_components/iron-icon/iron-icon.html">
</head>
```

The `<link rel="import" href="/bower_components/iron-icon/iron-icon.html">` is loaded in last position now, hence all dependencies are available at this point.

Works like a charm for me now.

**Comment by emmanuelbuah** [comment]
Thanks for the work around. Hopefully, a formal fix will get pushed out.

**Comment by mattiLeBlanc** [comment]
Hi guys, great work at debugging this.
I also have another workaround if you are interested, which doesnt force you to the modify the polymer objects:
Wait for the WebComponentsReady event to fire before you start rendering.

```
$(window).on "WebComponentsReady", (e)->
   window.WebComponentsReady =  true
```

I

**Comment by jurgenfink** [comment]
@mattiLeBlanc  Thank you for your input - very nice. Let's see how the Polymer Team will incorporate this issue in their update. At least, our findings show that the issue is related to the timing of availability of dependencies at run time (i.e. first loading time). 

Also a big thank you to @rohanray , @jatsrt, @sjmiles and @emmanuelbuah for having raised this issue here on Github and on Stack Overflow and for having tested certain approaches that we have come up with.

Any how, **Polymer rocks**, especially **1.0** - keep up the good work you folks at Polymer Team. Thank you.

And for non-_coffeescript_-lers, here the JavaScript output that Matti's code generates. I tested it and works like a charm, too (Chrome, FireFox, IE from Microbugs, Opera). Well done Matti :+1: 

```
    window.addEventListener('WebComponentsReady', function(e) {
        // do stuff here as components are loaded .....
    });
```

**Comment by mattiLeBlanc** [comment]
@jurgenfink My pleasure mate!
I did get on weird side effect with the WebComponentsReady solution in Blaze (using Meteor), that my whole scaffold's (paper-drawer-panel with header and tool bar) is broken. Instead of fullbleed, it is somewhere at the bottom all compressed. 
It looks like some of the child tags are not rendered within the content tags of outer polymer tags (you know that they have content area's using a select attribute to direct  the markup in certain area's).

I will have to test if this only happens with Blaze. However the icons load fine now..but not the rest :(

**Comment by mattiLeBlanc** [comment]
Hi Guys,

For me this issue only arises while using a template engine, in my case Blaze (Meteor.js).
When I render the icons without Blaze, they show up fine and dandy. I will also raise this issue on the Blaze git page.

**Comment by mattiLeBlanc** [comment]
@sjmiles has this been fixed or did you migrate it to another issue?

**Comment by ebidel** [comment]
Reopening until we confirm with @sjmiles. I'm getting bugged about this one.

**Comment by mattiLeBlanc** [comment]
I do see that here is a new iron-icons version 1.01. This one does FIX the loading issue for me.
I dont need WebComponentsReady event any more.

**Comment by emmanuelbuah** [comment]
@mattiLeBlanc I'm using blaze(meteor) as well and having the same issues you are having. The icons was first issue. I'm still digging around to figure out why rendering with a template throws everything out of place.

**Comment by ghost** [comment]
I am using Polymer in a Meteor 1.0 app. Everything is OK with Polymer 0.5.6, and I'm continuing to move my app forward with that. But with 'production-ready' Polymer 1.0 there are still some road blocks. This `[iron-icon::_updateIcon]` error _was_ one of them. I have just been able to extinguish these errors, which I have been seeing since at least 0.9, using the @jurgenfink remedy. Load order is definitely a factor, but the impact is inconsistent and counter-intuitive.

Initially, no icons would load as I got the error for each one on the page. When I used the 'WebComponentsReady' as a pre-render condition, only some of the error messages were extinguished. Oddly, some error messages that were associated with the same iconset as others that were eliminated still remained. For example, at one point, I had 6 errors associated with `could not find iconset 'social'`, yet after applying 'WebComponentsReady' as a pre-render condition 4 of these were eliminated and 2 remained. 

Per the @jurgenfink remedy, I went into the iconset .html (e.g. `social-icons.html`) file and commented out the `link` statement loading the `iron-icon.html` file and put an explicit `link` statement for the `iron-icon.html` file in my `head`, but well below the ones that load the iconsets. (Note: It was not necessary in my case to put explicit `link` statements for `iron-meta`, `iron-flex-layout`, nor `iron-iconset-svg`.) Once I did this for `iron-icons.html` and `social-icons.html`, the remaining errors disappeared. I could bring the errors back by uncommenting. Strangely enough I get errors for both `icons` and `social` by just uncommenting the `link` statement in the `iron-icons.html` file while leaving the comment in the `social-icons.html` file. Meanwhile, there are other iconsets, whose icons load fine without the need to comment out anything.

This is really weird.

BTW: I'm seeing this error on Chrome Version 44.0.2403.30 beta (64-bit), but not on Firefox 38.0.1, nor Safari 8.0.6 (all on OS X Yosemite).

**Comment by emmanuelbuah** [comment]
@dgtlife @mattiLeBlanc I was hoping we could work together to figure out our meteor + polymer woes here https://gitter.im/emmanuelbuah/MeteorPolymer#. We can post our findings here when we have an idea of whats causing the issue.

**Comment by emmanuelbuah** [comment]
This should probably be filed as a separate issue but I will go ahead and post it here since its related. I posted the below content relating to polymer & meteor issue https://github.com/meteor/meteor/issues/4528#issuecomment-110921223

Polymer components/tags inherits from dom element in order to participate in dom events: create, attach and detach. On create via document.createElement, the polymer elements generate specific children elements (per design). For example, document.createElement('paper-button') generates an elem with structure

```
<paper-button role="button" tabindex="0" aria-disabled="false" raised="" class="x-scope paper-button-0">
    <paper-ripple class="style-scope paper-button">
    <div id="background" class="style-scope paper-ripple"></div>
    <div id="waves" class="style-scope paper-ripple"></div>
  </paper-ripple>
    <paper-material animated="" class="content  style-scope paper-button x-scope paper-material-0" elevation="1">
     </paper-material>
</paper-button>
```

This happens in function - materializeTag in blaze/materalizer.js. After the element is created, blaze proceeds to append its children to the element. With our sample/example, the inner child of the button is the text BUTTON which results in

```
<paper-button role="button" tabindex="0" aria-disabled="false" raised="" class="x-scope paper-button-0">
    <paper-ripple class="style-scope paper-button">
    <div id="background" class="style-scope paper-ripple"></div>
    <div id="waves" class="style-scope paper-ripple"></div>
  </paper-ripple>
    <paper-material animated="" class="content  style-scope paper-button x-scope paper-material-0" elevation="1">
     </paper-material>
  BUTTON</paper-button>
```

Mean while, polymer expects a paper-button structure to look like this

```
<paper-button raised="" role="button" tabindex="0" aria-disabled="false" class="x-scope paper-button-0">
    <paper-ripple class="style-scope paper-button">
    <div id="background" class="style-scope paper-ripple"></div>
    <div id="waves" class="style-scope paper-ripple"></div>
  </paper-ripple>
    <paper-material animated="" class="content  style-scope paper-button x-scope paper-material-0" elevation="1">
      BUTTON
  </paper-material>
  </paper-button>
```

### Why its works on first load

On initial page load, the elements displays properly for when blaze materalizing kicks in for polymer element base library hasn't been loaded yet. Without the polymer library, the function materializeTag in blaze/materalizer.js produces the simple html element

```
<paper-button>BUTTON </paper-button>
```

and appends it to the dom. After polymer element library loads, it finds all polymer tags in the page and augments them to produce the final structure.

### Core of the issues

This issues will apply to any templating engine that renders/generates polymer elements via document.createElement('polymer element')

**Comment by emmanuelbuah** [comment]
@ebidel It looks like the the polymer team is aware and working on https://github.com/Polymer/polymer/blob/master/src/lib/experimental/patch-dom.html as a solution to polymer interoperability issues with existing dom manipulation lib's (templates libs). Can someone from your team offer insight as to how ready the patch is and how to go about using it.

**Comment by dplassgit** [comment]
I've given up on custom iconsets as a result of this bug. Built-ins seem to work though.

**Comment by polonski** [comment]
Polymerize https://github.com/loneleeandroo/meteor-polymerize seems to have solved this issue but it still doesn't work for me https://github.com/loneleeandroo/meteor-polymerize/issues/8

**Comment by lorensr** [comment]
My fix was adding 

```
  <link rel="import" href="/bower_components/iron-icons/iron-icons.html">
```

to the top of the list of component `<link>`s.

**Comment by vinyll** [comment]
Same issue here: `[iron-icon::_updateIcon]: could not find iconset 'icons', did you import the iconset?` on _Chromium 46_.
Works just fine on _Firefox Developer Edition 41_ and _Firefox 39_.

The @lorensr trick to put _iron-icons_ import at the very top did not work for me.

**Comment by ebidel** [comment]
Root issue is: https://github.com/PolymerElements/iron-icons/issues/18

**Comment by s-devaney** [comment]
Still suffering with this issue

**Comment by mattiLeBlanc** [comment]
uhm, maybe not something you want to hear, but I moved away from polymer in favour of Meteor+Angular with Angular Material (which rocks). No webcomponents so noone of the issues we are experiencing with Polymer and Blaze

**Comment by samccone** [comment]
:+1: just ran into this one

**Comment by ebidel** [comment]
This should be fixed with PolymerElements/iron-icons#18. PTAL.
